File: src/diagonal-traverse/diagonal-traverse.py
# https://leetcode.com/problems/diagonal-traverse

"""
3 x 3:
0,0; 0,1; 1,1; 2,1; 1,1; 0,2; 1,3; 2,1; 2,2;

when we step out of grid:
    if dr == up+right:
        - try moving column rigtward
        - elif try moving column downward
        - else done
    elif dr == down+left:
        - try moving row downward
        - else try moving column rightward


3 x 2:
"""


class Solution(object):
    def findDiagonalOrder(self, mat):
        n = len(mat)
        m = len(mat[0])
        i = 0
        j = 0
        end = (n - 1, m - 1)
        # 0: up+right, 1: down+left
        dr = 0
        arr = []
        while True:
            arr.append(mat[i][j])
            if dr == 0:
                right = j + 1 < m
                up = i - 1 >= 0
                down = i + 1 < n
                if right and up:
                    j += 1
                    i -= 1
                    continue
                dr = 1
                if right:
                    j += 1
                elif down:
                    i += 1
                elif (i, j) == end:
                    break
                else:
                    assert False, "bad up-right move"
            else:
                left = j - 1 >= 0
                down = i + 1 < n
                right = j + 1 < m
                if left and down:
                    j -= 1
                    i += 1
                    continue
                dr = 0
                if down:
                    i += 1
                elif right:
                    j += 1
                elif (i, j) == end:
                    break
                else:
                    assert False, "bad down-left move"
        return arr


if __name__ == "__main__":
    sol = Solution()
    tests = [
        ([[1, 2, 3], [4, 5, 6], [7, 8, 9]], [1, 2, 4, 7, 5, 3, 6, 8, 9]),
        (
            [
                [1, 2, 3, 4, 5],
                [6, 7, 8, 9, 10],
            ],
            [1, 2, 6, 7, 3, 4, 8, 9, 5, 10],
        ),
        ([[1, 2], [3, 4]], [1, 2, 3, 4]),
    ]
    for (mat, solution) in tests:
        result = sol.findDiagonalOrder(mat)
        assert result == solution, f"result {result} != solution {solution}"
    print("✅ All tests passed")

File: src/diagonal-traverse/diagonal-traverse.rs
// https://leetcode.com/problems/diagonal-traverse
struct Solution;

#[derive(Debug)]
enum Dir {
    UpRight,
    DownLeft,
}

impl Solution {
    pub fn find_diagonal_order(mat: Vec<Vec<i32>>) -> Vec<i32> {
        let n = mat.len();
        let m = mat[0].len();
        let mut arr = Vec::with_capacity(n * m);
        let mut dir = Dir::UpRight;
        let mut i = 0;
        let mut j = 0;

        loop {
            arr.push(mat[i][j]);
            println!("dir={:?}, i={}, j={}, arr={:?}", dir, i, j, arr);
            match (
                &dir,
                i >= 1,                   // up
                j + 1 < m,                // right
                i + 1 < n,                // down
                j >= 1,                   // left
                i == n - 1 && j == m - 1, // end
            ) {
                (_, _, _, _, _, true) => break,
                (Dir::UpRight, true, true, _, _, _) => {
                    j += 1;
                    i -= 1
                }
                (Dir::UpRight, false, true, _, _, _) => {
                    j += 1;
                    dir = Dir::DownLeft
                }
                (Dir::UpRight, false, false, true, _, _) => {
                    i += 1;
                    dir = Dir::DownLeft
                }
                (Dir::DownLeft, _, _, true, true, _) => {
                    j -= 1;
                    i += 1
                }
                (Dir::DownLeft, _, _, true, false, _) => {
                    i += 1;
                    dir = Dir::UpRight
                }
                (Dir::DownLeft, _, true, false, false, _) => {
                    j += 1;
                    dir = Dir::UpRight
                }
                (dir, up, rg, dw, lf, end) => {
                    panic!(
                        "bad move, dir={:?}, i={}, j={}, arr={:?}, up={}, rg={}, dw={}, lf={}, end={}",
                        dir, i, j, arr, up, rg, dw, lf, end
                    );
                }
            }
        }

        arr
    }
}

fn main() {
    let tests = vec![
        (
            vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]],
            vec![1, 2, 4, 7, 5, 3, 6, 8, 9],
        ),
        (
            vec![vec![1, 2, 3, 4, 5], vec![6, 7, 8, 9, 10]],
            vec![1, 2, 6, 7, 3, 4, 8, 9, 5, 10],
        ),
        (vec![vec![1, 2], vec![3, 4]], vec![1, 2, 3, 4]),
    ];


    for (mat, solution) in tests {
        let result = Solution::find_diagonal_order(mat);
        assert_eq!(result, solution);
    }
}
